# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [1]:
import pandas as pd

In [72]:
url1 = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df1 = pd.read_csv(url1)
url2 = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv"
df2 = pd.read_csv(url2)
url3 = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv"
df3 = pd.read_csv(url3)

In [73]:
df1.columns

Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Vehicle Class', 'Total Claim Amount'],
      dtype='object')

In [74]:
df2.columns

Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Total Claim Amount', 'Policy Type', 'Vehicle Class'],
      dtype='object')

In [75]:
df3.columns

Index(['Customer', 'State', 'Customer Lifetime Value', 'Education', 'Gender',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Total Claim Amount', 'Vehicle Class'],
      dtype='object')

In [76]:
df1.columns = df1.columns.str.lower().str.replace(" ", "_")
df1.columns

Index(['customer', 'st', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')

In [77]:
df1.columns = ['customer', 'state', 'gender', 'education', 'customer_lifetime_value', 'income', 'monthly_premium_auto', 'number_of_open_complaints', 'policy_type', 'vehicle_class', 'total_claim_amount']
df1.columns

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')

In [78]:
df2.columns = df2.columns.str.lower().str.replace(" ", "_")
df2.columns

Index(['customer', 'st', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'total_claim_amount', 'policy_type', 'vehicle_class'],
      dtype='object')

In [79]:
df2.columns = ['customer', 'state', 'gender', 'education', 'customer_lifetime_value', 'income', 'monthly_premium_auto', 'number_of_open_complaints', 'total_claim_amount', 'policy_type', 'vehicle_class']
df2.columns

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'total_claim_amount', 'policy_type', 'vehicle_class'],
      dtype='object')

In [80]:
df3.columns = df3.columns.str.lower().str.replace(" ", "_")
df3.columns

Index(['customer', 'state', 'customer_lifetime_value', 'education', 'gender',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'total_claim_amount', 'vehicle_class'],
      dtype='object')

In [81]:
print(df1.shape)
print(df2.shape)
print(df3.shape)

(4008, 11)
(996, 11)
(7070, 11)


In [82]:
print(df1.isnull().sum())
print(df2.isnull().sum())


customer                     2937
state                        2937
gender                       3054
education                    2937
customer_lifetime_value      2940
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64
customer                     0
state                        0
gender                       5
education                    0
customer_lifetime_value      4
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
total_claim_amount           0
policy_type                  0
vehicle_class                0
dtype: int64


In [83]:
print(df3.isnull().sum())


customer                     0
state                        0
customer_lifetime_value      0
education                    0
gender                       0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
total_claim_amount           0
vehicle_class                0
dtype: int64


In [84]:
df1.sample(20)

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
766,PM76175,Oregon,F,Bachelor,2114727.72%,49721.0,132.0,1/0/00,Corporate Auto,Sports Car,639.971388
3356,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3438,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2791,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
87,QC35222,California,NaN,Bachelor,268347.07%,48269.0,69.0,1/3/00,Corporate Auto,Four-Door Car,282.151207
3928,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
864,DK94262,Oregon,M,High School or Below,850712.88%,46754.0,106.0,1/1/00,Personal Auto,SUV,513.818403
2567,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3906,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3555,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [85]:
combined_df = pd.concat([df1, df2, df3], axis=0)
combined_df

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
...,...,...,...,...,...,...,...,...,...,...,...
7065,LA72316,California,M,Bachelor,23405.98798,71941.0,73.0,0,Personal Auto,Four-Door Car,198.234764
7066,PK87824,California,F,College,3096.511217,21604.0,79.0,0,Corporate Auto,Four-Door Car,379.200000
7067,TD14365,California,M,Bachelor,8163.890428,0.0,85.0,3,Corporate Auto,Four-Door Car,790.784983
7068,UP19263,California,M,College,7524.442436,21941.0,96.0,0,Personal Auto,Four-Door Car,691.200000


In [86]:
combined_df.isnull().sum()

customer                     2937
state                        2937
gender                       3059
education                    2937
customer_lifetime_value      2944
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64

In [87]:
combined_df.shape

(12074, 11)

In [88]:
combined_df = combined_df.dropna(how = "all")

In [89]:
combined_df.isnull().sum()

customer                       0
state                          0
gender                       122
education                      0
customer_lifetime_value        7
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64

In [90]:
combined_df.duplicated().sum()

np.int64(3)

In [91]:
combined_df = combined_df.drop_duplicates()

In [92]:
combined_df.duplicated().sum()

np.int64(0)

In [93]:
combined_df.reset_index(drop=True, inplace=True)

In [94]:
combined_df.duplicated(subset = ["customer"]).sum()

np.int64(78)

In [95]:
combined_df = combined_df.drop_duplicates(subset=["customer"])

In [96]:
combined_df.duplicated(subset = ["customer"]).sum()

np.int64(0)

In [97]:
combined_df.isnull().sum()

customer                       0
state                          0
gender                       122
education                      0
customer_lifetime_value        7
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64

In [98]:
combined_df["gender"] = combined_df["gender"].fillna("Unknown")

In [99]:
combined_df.isnull().sum()

customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      7
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64

In [100]:
combined_df[combined_df["customer_lifetime_value"].isna()]

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,RB50392,Washington,Unknown,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
78,SP81997,Washington,F,Master,NaN,41275.0,96.0,1/0/00,Personal Auto,Four-Door Car,41.122303
988,GT62080,Washington,M,High School or Below,NaN,55561.0,63.0,1/0/00,Personal Auto,Four-Door Car,227.872071
1393,RI52095,Washington,M,High School or Below,NaN,51878.0,66.0,1/1/00,Personal Auto,Four-Door Car,316.800000
1440,HF88410,Washington,Unknown,High School or Below,NaN,36765.0,66.0,1/1/00,Personal Auto,Four-Door Car,320.849072
1648,MW80236,Washington,M,Master,NaN,0.0,70.0,1/0/00,Personal Auto,Four-Door Car,336.000000
1929,TO60700,Washington,F,Bachelors,NaN,25859.0,74.0,1/0/00,Special Auto,Four-Door Car,355.200000


In [101]:
combined_df["customer_lifetime_value"] = pd.to_numeric(combined_df["customer_lifetime_value"].str.replace("%", "", regex=False), errors="coerce")

In [102]:
combined_df["customer_lifetime_value"].head()

0           NaN
1     697953.59
2    1288743.17
3     764586.18
4     536307.65
Name: customer_lifetime_value, dtype: float64

In [103]:
combined_df["customer_lifetime_value"].mean()

np.float64(780264.0244919786)

In [104]:
combined_df["customer_lifetime_value"].median()

572076.51

In [105]:
combined_df["customer_lifetime_value"] = combined_df["customer_lifetime_value"].fillna(combined_df["customer_lifetime_value"].median())


In [106]:
combined_df.isnull().sum()

customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64

In [107]:
combined_df

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,RB50392,Washington,Unknown,Master,572076.51,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
...,...,...,...,...,...,...,...,...,...,...,...
9129,LA72316,California,M,Bachelor,572076.51,71941.0,73.0,0,Personal Auto,Four-Door Car,198.234764
9130,PK87824,California,F,College,572076.51,21604.0,79.0,0,Corporate Auto,Four-Door Car,379.200000
9131,TD14365,California,M,Bachelor,572076.51,0.0,85.0,3,Corporate Auto,Four-Door Car,790.784983
9132,UP19263,California,M,College,572076.51,21941.0,96.0,0,Personal Auto,Four-Door Car,691.200000


In [108]:
combined_df = combined_df.reset_index()
combined_df

,index,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,0,RB50392,Washington,Unknown,Master,572076.51,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
...,...,...,...,...,...,...,...,...,...,...,...,...
9051,9129,LA72316,California,M,Bachelor,572076.51,71941.0,73.0,0,Personal Auto,Four-Door Car,198.234764
9052,9130,PK87824,California,F,College,572076.51,21604.0,79.0,0,Corporate Auto,Four-Door Car,379.200000
9053,9131,TD14365,California,M,Bachelor,572076.51,0.0,85.0,3,Corporate Auto,Four-Door Car,790.784983
9054,9132,UP19263,California,M,College,572076.51,21941.0,96.0,0,Personal Auto,Four-Door Car,691.200000


In [109]:
combined_df = combined_df.drop(columns=["level_0", "index"])

KeyError: "['level_0'] not found in axis"

In [110]:
combined_df

,index,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,0,RB50392,Washington,Unknown,Master,572076.51,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
...,...,...,...,...,...,...,...,...,...,...,...,...
9051,9129,LA72316,California,M,Bachelor,572076.51,71941.0,73.0,0,Personal Auto,Four-Door Car,198.234764
9052,9130,PK87824,California,F,College,572076.51,21604.0,79.0,0,Corporate Auto,Four-Door Car,379.200000
9053,9131,TD14365,California,M,Bachelor,572076.51,0.0,85.0,3,Corporate Auto,Four-Door Car,790.784983
9054,9132,UP19263,California,M,College,572076.51,21941.0,96.0,0,Personal Auto,Four-Door Car,691.200000


In [111]:
combined_df["gender"].unique()

array(['Unknown', 'F', 'M', 'Femal', 'Male', 'female'], dtype=object)

In [112]:
combined_df["gender"] = combined_df["gender"].str.strip().str.lower().replace({"female": "F", "femal": "F", "f": "F", "male": "M", "m": "M"})
combined_df["gender"].unique()

array(['unknown', 'F', 'M'], dtype=object)

In [113]:
combined_df["state"].unique()

array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', 'Cali',
       'AZ', 'WA'], dtype=object)

In [114]:
combined_df["state"] = combined_df["state"].replace({"AZ": "Arizona", "WA": "Washington", "Cali": "California"})

In [115]:
combined_df["state"].unique()

array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon'],
      dtype=object)

In [116]:
combined_df["education"].unique()

array(['Master', 'Bachelor', 'High School or Below', 'College',
       'Bachelors', 'Doctor'], dtype=object)

In [117]:
combined_df["education"] = combined_df["education"].replace({"Bachelors": "Bachelor"})

In [118]:
combined_df["education"].unique()

array(['Master', 'Bachelor', 'High School or Below', 'College', 'Doctor'],
      dtype=object)

In [119]:
combined_df["vehicle_class"] = combined_df["vehicle_class"].replace({"Luxury SUV": "Luxury", "Luxury Car": "Luxury", "Sports Car": "Luxury"})

In [120]:
combined_df["vehicle_class"].unique()

array(['Four-Door Car', 'Two-Door Car', 'SUV', 'Luxury'], dtype=object)

In [122]:
combined_df

,index,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,0,RB50392,Washington,unknown,Master,572076.51,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
...,...,...,...,...,...,...,...,...,...,...,...,...
9051,9129,LA72316,California,M,Bachelor,572076.51,71941.0,73.0,0,Personal Auto,Four-Door Car,198.234764
9052,9130,PK87824,California,F,College,572076.51,21604.0,79.0,0,Corporate Auto,Four-Door Car,379.200000
9053,9131,TD14365,California,M,Bachelor,572076.51,0.0,85.0,3,Corporate Auto,Four-Door Car,790.784983
9054,9132,UP19263,California,M,College,572076.51,21941.0,96.0,0,Personal Auto,Four-Door Car,691.200000


In [125]:
combined_df.isnull().sum()

index                        0
customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64

In [126]:
def clean_complaints(value):
    value = str(value)
    
    if "/" in value:
        return value.split("/")[1]
    else:
        return value

combined_df["number_of_open_complaints"] = (
    combined_df["number_of_open_complaints"]
    .apply(clean_complaints)
)

combined_df["number_of_open_complaints"] = pd.to_numeric(
    combined_df["number_of_open_complaints"],
    errors="coerce"
)

In [124]:
combined_df["number_of_open_complaints"].unique()

array([0, 2, 1, 3, 5, 4])

In [127]:
combined_df

,index,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,0,RB50392,Washington,unknown,Master,572076.51,0.0,1000.0,0,Personal Auto,Four-Door Car,2.704934
1,1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,0,Personal Auto,Four-Door Car,1131.464935
2,2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,0,Personal Auto,Two-Door Car,566.472247
3,3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,0,Corporate Auto,SUV,529.881344
4,4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,0,Personal Auto,Four-Door Car,17.269323
...,...,...,...,...,...,...,...,...,...,...,...,...
9051,9129,LA72316,California,M,Bachelor,572076.51,71941.0,73.0,0,Personal Auto,Four-Door Car,198.234764
9052,9130,PK87824,California,F,College,572076.51,21604.0,79.0,0,Corporate Auto,Four-Door Car,379.200000
9053,9131,TD14365,California,M,Bachelor,572076.51,0.0,85.0,3,Corporate Auto,Four-Door Car,790.784983
9054,9132,UP19263,California,M,College,572076.51,21941.0,96.0,0,Personal Auto,Four-Door Car,691.200000


# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [1]:
import pandas as pd

In [2]:
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv"

df_mkt = pd.read_csv(url)

In [15]:
pd.set_option('display.max_columns', None)

In [16]:
df_mkt

,unnamed:_0,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,income,location_code,marital_status,monthly_premium_auto,months_since_last_claim,months_since_policy_inception,number_of_open_complaints,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
0,0,DK49336,Arizona,4809.216960,No,Basic,College,2011-02-18,Employed,M,48029,Suburban,Married,61,7.000000,52,0.000000,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,A,2
1,1,KX64629,California,2228.525238,No,Basic,College,2011-01-18,Unemployed,F,0,Suburban,Single,64,3.000000,26,0.000000,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,A,1
2,2,LZ68649,Washington,14947.917300,No,Basic,Bachelor,2011-02-10,Employed,M,22139,Suburban,Single,100,34.000000,31,0.000000,2,Personal Auto,Personal L3,Offer3,Call Center,480.000000,SUV,Medsize,A,2
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,2011-01-11,Employed,M,49078,Suburban,Single,97,10.000000,3,0.000000,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A,1
4,4,QA50777,Oregon,9025.067525,No,Premium,Bachelor,2011-01-17,Medical Leave,F,23675,Suburban,Married,117,15.149071,31,0.384256,7,Personal Auto,Personal L2,Offer1,Branch,707.925645,Four-Door Car,Medsize,A,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10905,10905,FE99816,Nevada,15563.369440,No,Premium,Bachelor,2011-01-19,Unemployed,F,0,Suburban,Married,253,15.149071,40,0.384256,7,Personal Auto,Personal L1,Offer3,Web,1214.400000,Luxury Car,Medsize,A,1
10906,10906,KX53892,Oregon,5259.444853,No,Basic,College,2011-01-06,Employed,F,61146,Urban,Married,65,7.000000,68,0.000000,6,Personal Auto,Personal L3,Offer2,Branch,273.018929,Four-Door Car,Medsize,A,1
10907,10907,TL39050,Arizona,23893.304100,No,Extended,Bachelor,2011-02-06,Employed,F,39837,Rural,Married,201,11.000000,63,0.000000,2,Corporate Auto,Corporate L3,Offer1,Web,381.306996,Luxury SUV,Medsize,A,2
10908,10908,WA60547,California,11971.977650,No,Premium,College,2011-02-13,Employed,F,64195,Urban,Divorced,158,0.000000,27,4.000000,6,Personal Auto,Personal L1,Offer1,Branch,618.288849,SUV,Medsize,A,2


In [4]:
df_mkt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 27 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   unnamed:_0                     10910 non-null  int64  
 1   customer                       10910 non-null  object 
 2   state                          10910 non-null  object 
 3   customer_lifetime_value        10910 non-null  float64
 4   response                       10910 non-null  object 
 5   coverage                       10910 non-null  object 
 6   education                      10910 non-null  object 
 7   effective_to_date              10910 non-null  object 
 8   employmentstatus               10910 non-null  object 
 9   gender                         10910 non-null  object 
 10  income                         10910 non-null  int64  
 11  location_code                  10910 non-null  object 
 12  marital_status                 10910 non-null 

In [5]:
df_mkt.isnull().sum()

unnamed:_0                       0
customer                         0
state                            0
customer_lifetime_value          0
response                         0
coverage                         0
education                        0
effective_to_date                0
employmentstatus                 0
gender                           0
income                           0
location_code                    0
marital_status                   0
monthly_premium_auto             0
months_since_last_claim          0
months_since_policy_inception    0
number_of_open_complaints        0
number_of_policies               0
policy_type                      0
policy                           0
renew_offer_type                 0
sales_channel                    0
total_claim_amount               0
vehicle_class                    0
vehicle_size                     0
vehicle_type                     0
month                            0
dtype: int64

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [18]:
df_mkt["sales_channel"].unique()

array(['Agent', 'Call Center', 'Branch', 'Web'], dtype=object)

In [17]:
pivot_df = pd.pivot_table(df_mkt, index="sales_channel", values="total_claim_amount", aggfunc="sum").round(2)
pivot_df

,total_claim_amount
sales_channel,
Agent,1810226.82
Branch,1301204.00
Call Center,926600.82
Web,706600.04


In [20]:
pivot_df["Percentage"] = (pivot_df["total_claim_amount"] / pivot_df["total_claim_amount"].sum() * 100).round(2)
pivot_df

,total_claim_amount,Percentage
sales_channel,,
Agent,1810226.82,38.15
Branch,1301204.00,27.42
Call Center,926600.82,19.53
Web,706600.04,14.89


Firstly, I would like to note that the exercise asks for mail as a sales channel. However, when checking the options in sales channel, mail is not one. The channels used were agent, branch, call center and web.

By looking at the pivot table we can easily see that the agent channel was the most effective, 38%, followed not by far by branch.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [27]:
pivot_lifetime = pd.pivot_table(df_mkt, index="gender", columns="education", values=["customer_lifetime_value"], aggfunc="mean").round(2)
pivot_lifetime

customer_lifetime_value                                         \
education                Bachelor  College   Doctor High School or Below   
gender                                                                     
F                         7874.27  7748.82  7328.51              8675.22   
M                         7703.60  8052.46  7415.33              8149.69   

                    
education   Master  
gender              
F          8157.05  
M          8168.83

The pivot shows the average customer lifetime value (CLV) varies slightly across education and gender. Females with High School or below have the highest clv, 8675.22, while male customer with a Master have the highest clv , 8168.83. Difference between gender is small, which means that gender does not impact a lot on clv.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [ ]:
# Your code goes here